url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/Keyboard.java

```diff
@@ -0,0 +1,619 @@
+  public static final int USER_FIRE = 4;
+
+  private static final int[] ARROWS = new int[] {
+    KeyEvent.VK_UP, 0, KeyEvent.VK_DOWN, 0,
+    KeyEvent.VK_LEFT, 0, KeyEvent.VK_RIGHT, 0,
+    KeyEvent.VK_CONTROL, KeyEvent.KEY_LOCATION_RIGHT};
+
+  private int[] userDefinedStick = ARROWS;
+
+  private CIA cia;
+  private C64Screen screen;
+
+  private void remap(char key, char newkey) {
+    for (int i = 0, n = keytableDef.length; i < n; i++) {
+      if (keytableDef[i][0] == key) {
+        keytableDef[i][0] = newkey;
+        System.out.println("Remapped: " + (char) key + " => " + (char) newkey +
+            "  key " + (int) key + " => " + (int) newkey);
+        return;
+      }
+    }
+  }
+
+
+  private boolean doMap(String ltarget) {
+    for (int i = 0, n = KEYMAPS.length; i < n; i++) {
+      char[] map = KEYMAPS[i];
+      String lang = "" + map[0] + map[1];
+      System.out.println("Checking map for: " + lang);
+      if (ltarget.equals(lang)) {
+        System.out.println("Found map - mapping...");
+        for (int j = 2, m = map.length; j < m; j += 2) {
+          remap(map[j], map[j + 1]);
+        }
+        // Finished!
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Creates a new <code>Keyboard</code> instance.
+   *
+   */
+  public Keyboard(C64Screen screen, CIA cia, int[] memory) {
+
+    Locale locale = Locale.getDefault();
+    System.out.println("Locale: " + locale);
+
+    String lang = locale.getLanguage();
+    if (!doMap(lang)) {
+      System.out.println("Could not find map for keyboard: " + lang);
+    }
+
+    this.cia = cia;
+    this.screen = screen;
+
+    for (int i = 0; i < keytable.length; i++) {
+      keytable[i][0] = -1;
+    }
+
+    for (int i = 0; i < keytableDef.length; i++) {
+      int key = keytableDef[i][0];
+      keytable[key][0] = keytableDef[i][1];
+      keytable[key][1] = keytableDef[i][2];
+      keytable[key][2] = keytableDef[i][3];
+    }
+
+    reset();
+  }
+
+  public void registerHotKey(int key, int modflag, String script, Object o) {
+    if (hotkeyScript == null) {
+      c64script = new C64Script();
+      hotkeyScript = new ArrayList();
+    }
+    hotkeyScript.add(new Object[] {script, o, new int[] {key, modflag}});
+  }
+
+  public void setStick(boolean one) {
+    if (one)
+      stick = IO_OFFSET + 0xdc01;
+    else
+      stick = IO_OFFSET + 0xdc00;
+  }
+
+  private int getUserStick(int key, int location) {
+    if (userDefinedStick != null) {
+      for (int i = 0, n = userDefinedStick.length; i < n; i += 2) {
+        if (key == userDefinedStick[i]) {
+          int cmpLoc = userDefinedStick[i + 1];
+          if (cmpLoc == 0 || location == userDefinedStick[i + 1]) {
+//          System.out.println("Joystick emulation trigger");
+            return i / 2;
+          }
+        }
+      }
+    }
+    return -1;
+  }
+
+  // # Chars in char buffer - 0xc6 - need not to be set here...
+  public void keyPressed(KeyEvent event) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 22. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 310-322 into a separate method.